### ISSUE TITLE
NET-1641 Fix S1905 FP: casting `IEnumerable<string?>` to `IEnumerable<string>`

### ISSUE THREAD

**Issue by fiotti** [SonarSource/sonar-dotnet#8413]
### Description

False positive S1905: casting `IEnumerable<string?>` to `IEnumerable<string>`.

### Repro steps

```csharp
class SonarTest
{
    public IEnumerable<string> GetNonNullStrings(IEnumerable<string?> strings)
    {
        // S1905: Remove this unnecessary cast to 'IEnumerable<string>'.
        return (IEnumerable<string>)strings.Where(s => s != null);
    }
}
```

### Expected behavior

I didn't expect any diagnostics to be emitted.

### Actual behavior

False positive.

### Known workarounds

```csharp
class SonarTest
{
    public IEnumerable<string> GetNonNullStrings(IEnumerable<string?> strings)
    {
        return strings.OfType<string>();
    }
}
```

### Related information

* SonarLint for Visual Studio 2022, version 7.4.0.80741
* Visual Studio Professional 2022 (64 bit), version 17.7.6
* dotnet 7.0.403
* MSBuild version 17.7.2+d6990bcfa for .NET Framework, version 17.7.2.37605
* Windows 11 Pro 22H2 22621.2715

**Comment by sebastien-marichal** [comment]
Hello @fiotti,

Thank you for reporting this issue!

I confirm this as an FP.
I will add a reproducer to our code base to tackle it in the future!

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1641](https://sonarsource.atlassian.net/browse/NET-1641)

[NET-1641]: https://sonarsource.atlassian.net/browse/NET-1641?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by doterik** [comment]
###### yac (yet another comment)

When I remove this cast:  `var buffer = (Span<char>)stackalloc char[bufferSize];`, as warned by s1905,

I end up with a `char*`, which requires an unsafe context—manageable.

However, calling `PInvoke.GetVolumePathName(filepath, buffer)` with that `char*` doesn't compile. That's a dealbreaker.

**Comment by martin-strecker-sonarsource** [comment]
Thank you @doterik This is another issue and I created internal ticket [NET-2197](https://sonarsource.atlassian.net/browse/NET-2197) and will work on a fix.

[NET-2197]: https://sonarsource.atlassian.net/browse/NET-2197?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by martin-strecker-sonarsource** [comment]
The `IEnumerable<string?>` case and the `stackalloc` case have both been fixed in version [10.15.0.120848](https://github.com/SonarSource/sonar-dotnet/releases/tag/10.15.0.120848)
